plugins/nf-amazon/src/main/com/upplication/s3fs/ng/S3ParallelDownload.java

```diff
@@ -0,0 +1,268 @@
+        return new SequenceInputStream(
+                getDownloadIterator(
+                        prepareGetPartRequests(bucketName, key)
+                )
+        );
+    }
+
+    private Task submitTask(GetObjectRequest request) {
+        return new Task(request, executor.submit(downloadChunk(request)));
+    }
+
+    private Enumeration<InputStream> getDownloadIterator(Iterator<GetObjectRequest> parts) {
+
+        // FIFO queue of parts downloading in parallel
+        Queue<Task> tasks = new LinkedList<>();
+
+        // Add at least one part to download
+        if (parts.hasNext()) {
+            tasks.add(submitTask(parts.next()));
+        }
+
+        // Add up to 'numWorkers' parts if the executor is idle
+        int submitted = 1;
+        while (parts.hasNext() && submitted < opts.numWorkers() && executor.getActiveCount() < opts.numWorkers()) {
+            tasks.add(submitTask(parts.next()));
+            submitted++;
+        }
+
+        return new Enumeration<InputStream>() {
+
+            public boolean hasMoreElements() {
+                return !tasks.isEmpty() || parts.hasNext();
+            }
+
+            public InputStream nextElement() {
+
+                // Add next part to download
+                if (parts.hasNext()) {
+                    tasks.add(submitTask(parts.next()));
+                }
+
+                // Check if the executor is idle and submit one extra work to slowly increase the load
+                if (executor.getActiveCount() < opts.numWorkers() && parts.hasNext()) {
+                    tasks.add(submitTask(parts.next()));
+                }
+
+                if (tasks.isEmpty()) {
+                    throw new NoSuchElementException();
+                }
+
+                Task task = tasks.poll();
+                try {
+                    return task.future.get();
+                } catch (Throwable e) {
+                    Throwable cause = e.getCause();
+                    if (cause instanceof SdkClientException && !Thread.currentThread().isInterrupted()) {
+                        int size = Math.max(executor.getMaximumPoolSize() - 1, 4);
+                        executor.setCorePoolSize(size);
+                        executor.setMaximumPoolSize(size);
+                        log.warn("{}. Reducing executor pool size to {} current active threads {}", cause.getMessage(), size, executor.getActiveCount());
+                        try {
+                            Thread.sleep(1000);
+                        } catch (InterruptedException ex) {
+                            cleanUpAfterException();
+                            Thread.currentThread().interrupt();
+                        }
+                        log.warn("Downloading at main thread chunk {}; path={}.", getLogRef(task.request), getLogPath(task.request));
+                        return new DownloadOnDemandInputStream(task.request, s3Client);
+                    }
+
```

Comment:
Is this really needed? The `pool` should not thrown ant exception

Reply:
It is the download who is throwing the exception. This one comes from the `s3client.getObject` call. I'm handling it here to be able to reduce the pool max threads.

Reply:
I see, then we need to add it again because I've removed it